---
params:
  root: "../data"
  pidn: 90
---

```{r setup, include=F}
library(tidyverse)
source("../R/mutt.R")
knitr::opts_chunk$set(echo = F)
do.call(Sys.setenv, params)
```

Download source feature metadata from NCBI.

```{bash}
[[ -f "$root/blast.tsv" ]] && awk '/^[^#]/ { print $2; }' "$root/blast.tsv" | uniq > "$root/acc.txt"

{ 
  printf "accver\ttaxid\ttitle\tsubtype\tsubname\n"; 
  ../py/eutil.py "$root/acc.txt" -eutil esummary -param retmode=json | \
    tee "$root/src.json" | \
    jq -r '.result | del(.uids) | map([.accessionversion, .taxid, .title, .subtype, .subname] | @tsv) | .[]' 
} | \
  ../R/subtype.R - subtype subname 2> /dev/null | \
  ../R/lubridate.R - collection_date 2> /dev/null > "$root/src.tsv"
```

Download sequences from NCBI.

```{bash}
n="$(head -n 1 "$root/src.tsv" | tr '\t' '\n' | cat -n | grep collection_date | cut -f 1 | tr -d ' ')"

cut -f 1,2,"$n" "$root/src.tsv" | \
  awk 'NR > 1 && $3 != "NA"' | \
  sort -k 3 | \
  tee >(awk '{ printf "s/(%s.*)/\\1|%s|%d/\n", $1, $3, $2; }' > "$root/def.sed") | \
  cut -f 1 | \
  ../py/eutil.py - -param rettype=fasta > "$root/lib.fna"
```

Perform global/local alignment.

```{bash}
glsearch36 -m 8CB -T 8 "$root/qry.fna" "$root/lib.fna" > "$root/aln.tsv"
```

Extract subject regions with identity greater than or equal to the threshold.

```{bash}
rm -f "$root/lib.fna.fai"
awk -v id="$pidn" '/^[^#]/ && $3 >= id { printf "%s:%d-%d\n", $2, $9, $10; }' "$root/aln.tsv" | \
  samtools faidx "$root/lib.fna" -r - | \
  sed -E -f "$root/def.sed" > "$root/reg.fna"
```

Perform multiple alignment.

```{bash}
mafft --auto --adjustdirection --thread -1 "$root/reg.fna" > "$root/msa.fna" 2> "$root/msa.log"
```

...

```{r, fig.width=10, fig.height=4}
msa <- ape::read.dna(file.path(params$root, "msa.fna"), format = "fasta", as.character = T) %>% toupper()
msa <- msa[2:nrow(msa), ]
unq <- apply(msa, 2, unique)

df.snp <- call_snp(msa)
df.ind <- call_ind(msa)

ggplot() +
  geom_point(data = df.snp, aes(pos, accver, shape = call, color = call), size = 8) +
  geom_point(data = df.ind, aes(pos, accver, shape = call), size = 4, position = position_nudge(y = 0.25)) +
  scale_shape_manual(values = setNames(
    c(124, 124, 124, 124, 6, 2), 
    c("trs", "trv", "sim", "dis", "ins", "del")
  )) +
  facet_grid(taxid ~ ., scales = "free_y") +
  xlim(1, ncol(msa)) + 
  theme_minimal() + 
  theme(legend.position = "bottom")
```

Infer tree...

```{bash}
rm -f "$root/phy."*
iqtree -s "$root/msa.fna" -pre "$root/phy" -alrt 1000 -bb 1000 -bnni -nt AUTO
```

