```{r}
# variables
root <- "exp"
subroot <- "HAdV-4"
```

```{bash, script}
mkdir -p exp/HAdV-4 && cd exp/HAdV-4

# efetch
[ ! -f rec.fna ] && awk 'NR > 1 { print $1; }' meta.tsv | tr '\n' ',' | xargs efetch -db nuccore -format fasta -id > rec.fna

# meta
awk -F '\t' 'NR > 1 { printf "/^>/ s/%s/%s_%s_%s/\n", $1, $1, $4, $5; }' meta.tsv > meta.sed

# mafft
[ ! -f msa-1.fna ] && 
sed '/^>/ s/ .*//' rec.fna | sed -f meta.sed | mafft --auto --adjustdirection --thread 8 - > msa-1.fna 2> msa-1.log

# gubbins
sed '/^[^>]/ s/[^acgtACGT-]/N/g' msa-1.fna > msa-2.fna
[ ! -f gub.log ] && run_gubbins.py --prefix gub --iterations 1000 --threads 8 msa-2.fna > gub.log
bedtools maskfasta -fi msa-1.fna -fo msa-2.fna -bed gub.recombination_predictions.gff

# iqtree
[ ! -f phy.log ] && iqtree -s msa-2.fna -pre phy -alrt 1000 -bb 1000 -bnni -nt 8 > /dev/null

# beautify
line=( $(
  grep -v WARNING phy.log | awk '{{$1=$1;}} //1;' | \
    grep -A 131 '1 JC' | grep -v +R | sort -n -k 7 | head -n 1 | \
    sed -e 's/+G4/+G/g' -e 's/+F//g'
) );
for clock in "rex" "rln" "str"; do
  for coal in "con" "exp"; do
    ../../../../workflow/scripts/beautify.py \
      -dregex "_(\d{4})_" -dformat "%Y" -len_mcmc 100000000 -stem "$clock-$coal" \
      msa-2.fna ../../../../workflow/templates "${line[1]}" "$clock" "$coal" > "$clock-$coal.xml"
  done
done
```

```{r}
msa <- 
  file.path(root, subroot, "msa-1.fna") %>%
  ape::read.dna(format = "fasta", as.character = T) %>%
  toupper()

tree <- file.path(root, subroot, "gub.final_tree.tre") %>% ape::read.tree()
tree <- ape::ladderize(tree)
lvls <- tree$tip.label[tree$edge[tree$edge[,2] <= length(tree$tip.label), 2]]

vcf <- 
  bind_rows(merge(call_snp(msa), muts), mutate(call_ind(msa), pos = start)) %>%
  mutate(
    accver = factor(rownames(msa)[idx], levels = lvls), 
    call = factor(call, levels = c("trs", "trv", "sim", "dis", "ins", "del"))
  )

len <- ncol(msa)
colors <- list(trs = "blue", trv = "red", sim = "purple", dis = "orange", ins = "black", del = "black")
shapes <- list(trs = 124, trv = 124, sim = 124, dis = 124, ins = 6, del = 2)

recom <-
  file.path(root, subroot, "gub.recombination_predictions.gff") %>%
  ape::read.gff() %>%
  apply(1, function(row) {
    read_delim(str_split(row["attributes"], ";")[[1]], "=", col_names = c("key", "val"), col_types = "cc") %>%
      filter(key == "taxa") %>% pull(val) %>% str_trim() %>% str_split("\\s+") %>% unlist() %>%
      enframe(name = NULL, value = "accver") %>%
      mutate(start = row["start"], end = row["end"])
  }) %>%
  bind_rows() %>%
  mutate_at("accver", factor, levels = lvls) %>%
  mutate_at(c("start", "end"), as.integer)
```

```{r}
ggplot() +
  geom_point(data = vcf, aes(pos, accver, color = call, shape = call)) +
  geom_rect(
    data = recom,
    aes(
      xmin = start - 0.5, xmax = end + 0.5, 
      ymin = as.numeric(accver) - 0.5, ymax = as.numeric(accver) + 0.5
    ), 
    fill = "lightgrey", color = "black", alpha = 0.25
  ) +
  scale_y_discrete(limits = lvls) +
  scale_color_manual(values = colors) +
  scale_shape_manual(values = shapes) +
  xlim(1, len) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    text = element_text(family = "mono"),
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    axis.text.y = element_text(hjust = 0)
  )
```

```{r}
df.bf <-
  file.path(root, subroot) %>%
  list.files(pattern = ".mle.result.log", recursive = T, full.names = T) %>%
  lapply(parse_mle_report) %>%
  bind_rows() %>%
  mutate(acc = basename(dirname(path)), name = basename(path)) %>%
  separate(name, c("clock", "coalescent"), sep = "[-.]", extra = "drop") %>%
  group_by(acc) %>%
	mutate(
	  `BF (PS)` = PS - PS[which(clock == "str" & coalescent == "con")],
		`BF (SS)` = SS - SS[which(clock == "str" & coalescent == "con")]
	) %>%
  ungroup() %>%
  arrange(acc, desc(`BF (PS)`))

logs <-
  file.path(root, subroot) %>%
  list.files(pattern = "\\w+-\\w+.log", recursive = T, full.names = T) %>%
  grep("(rex|rln|str)-(con|exp).log", ., value = T)

pburn <- 0.1

df.ess <-
  mclapply(logs, function(path) {
    df <- parse_beast_log(path)
    interval <- df$state[2] - df$state[1]
    select(df, joint, prior, likelihood, age.root., meanRate, coalescent) %>%
      apply(2, function(ele) calc_ess(remove_burn_in(ele, pburn), interval)) %>%
      c(path = path)
  }) %>%
  map_df(bind_rows) %>%
  mutate(path = basename(path)) %>%
  select(path, everything()) %>%
  separate(path, c("clock", "coalescent"), sep = "[-.]", extra = "drop") %>%
  mutate_at(3:ncol(.), as.double)

df.est <-
  mclapply(logs, function(path) {
    parse_beast_log(path) %>%
      apply(2, median) %>%
      c(path = path)
  }) %>%
  map_df(bind_rows) %>%
  mutate(path = basename(path)) %>%
  select(path, everything()) %>%
  separate(path, c("clock", "coalescent"), sep = "[-.]", extra = "drop") %>%
  mutate_at(4:ncol(.), as.double)

df.res <-
  filter(df.ess, joint >= 200, prior >= 200, likelihood >= 200) %>%
  select(clock, coalescent) %>%
  merge(df.bf) %>%
  merge(df.est) %>%
  select(clock, coalescent, `BF (PS)`, age.root., meanRate) %>%
  arrange(desc(`BF (PS)`))
```

```{r, warning=FALSE}
tree <-
  with(df.res[1,], paste0(clock, "-", coalescent, ".mcc.tree")) %>%
  file.path(root, subroot, .) %>%
  read.beast()

tip.date <- str_match(tree@phylo$tip.label, "_(\\d{4})_")[, 2]
ggtree(tree, mrsd = paste0(min(tip.date), "-01-01")) +
  geom_tiplab(linesize = 1, align = T, color = "black") +
  geom_range("height_0.95_HPD") +
  theme_tree2() +
  theme(
    legend.position = "bottom",
    panel.grid.major.x = element_line(color="black", size = .25),
    panel.grid.minor.x = element_line(color="grey", size = .25)
  )
```

```{r}
knitr::kable(df.res)
```

