---
title: "Molecular Clock analysis of <i>Human adenovirus</i>"
date: "`r format(Sys.time(), '%B %d, %Y')`"
author: "Daniel Antonio Negrón"
institution: "George Mason University"
division: "College of Science"
advisor: "Don Seto, PhD"
altadvisor: "Patrick Gillevet, PhD & Sterling Thomas, PhD"
department: "School of Systems Biology"
degree: "Doctor of Philosophy"
bibliography: "references.bib"
csl: "csl/apa.csl"
site: bookdown::bookdown_site
documentclass: book
output:
  bookdown::gitbook: default
---

```{r setup, include=FALSE}
# packages
library(tidyverse)
library(lubricity)
library(lubridate)
library(ggtree)
library(tidytree)
library(BactDating)
library(parallel)
library(tracerer)
library(treeio)
library(kableExtra)
# scripts
source("../scripts/util.R")
# options
knitr::opts_chunk$set(
  echo = F, 
  fig.width = 10, fig.height = 10, fig.fullwidth = TRUE, fig.align = "center"
)
# variables
root <- "exp"
```

# Committee {.unnumbered}

George Mason University  
College of Science  
School of Systems Biology  
PhD Bioinformatics and Computational Biology  

Members:  
- Don Seto, PhD (Chair)  
- Patrick Gillevet, PhD  
- Sterling Thomas, PhD  

# Abstract

Members of the *Mastadenovirus* genus include re-emergent, globally-distributed pathogens with mammalian hosts. Advances in sequencing technology have yielded large, publicly-available data sets that are useful for understanding the genotypes that infect humans. The combination of DNA sequence, isolation date, and location help reconstruct the trajectory of different strains during outbreaks and is crucial for enhancing countermeasure quality. Accordingly, molecular clock techniques cluster samples and estimate divergence times and substitution rates. Sample data for such analysis is difficult to assemble and sanitize for large data sets. Thus, the proposed research includes the development of a protocol and automated pipeline that constructs data sets for molecular clock analysis. Preliminary work applied the protocol to the study of genotypes E4 and B55.

# Introduction

This preliminary research relied on molecular clock methods to estimate substitution rates and divergence times in *Mastadenovirus* samples. Such analysis requires a highly curated data set of time-stamped, homologous nucleic acid sequences representative of a measurably evolving population [@drummondMeasurablyEvolvingPopulations2003]. Thus, a preprocessing workflow is necessary to extract and normalize sampling date information. The presence of heterogeneous or ambiguous date formats is a complicating factor and a time sink. Also, manual processing becomes intractable as the number of samples increases and potentially generates costly downstream errors. As a result, this research defines a generic workflow and automated, parallel pipeline that coordinates the reliable execution of preprocessing and analytical tasks in a reproducible manner. This is a preliminary molecular clock study that replicates an HAdV-E4 paper [@gonzalezGenomicCharacterizationHuman2019] and presents the automated characterization of HAdV-B55.

## Significance

*Human adenovirus* is a diverse pathogen with global distribution that causes sporadic outbreaks. In 1953, the first known epidemic occurred at the Fort Leonard Wood U.S. Army installation in Missouri. A patient in this large outbreak presented with pneumonia-like symptoms and provided a throat wash sample that contained the first viral isolate, initially called “adenoid degeneration agent” and later adenovirus [@hillemanRecoveryNewAgent1954; @roweIsolationCytopathogenicAgent1953]. Subsequent outbreaks resulted in the discovery and characterization of the *Adenoviridae* family.

Data acquired from various outbreaks have yielded data sets suitable for molecular clock analysis. This preliminary research characterizes molecular evolution by testing clock and population models, calculating substitution rates, and estimating time to most recent common ancestor (TMRCA). This data helps public health officials make informed decision with respect to biosurveillance efforts and vaccine development for the general population. In addition, these methods are applicable to other organisms.

### HAdV-E4

Type 4 has the distinction of being the first recorded adenovirus isolate [@hillemanRecoveryNewAgent1954; @roweIsolationCytopathogenicAgent1953; @topControlAdenovirusAcute1975]. It is also the only member of the E species [@gonzalezGenomicCharacterizationHuman2019]. Evidence also suggests a zoonotic origin with similarity to Simian adenovirus 26 [@dehghanComputationalAnalysisFour2013; @jacobsCharacterizationManipulationHuman2004]. Since 1953, multiple outbreaks have afflicted both military and civilian populations with fatal outcomes [@arigaNewGenomeType2004; @artiedaSwimmingPoolrelatedOutbreak2009; @centersfordiseasecontrolandpreventioncdcTwoFatalCases2001; @dangeloPharyngoconjunctivalFeverCaused1979; @hendrixLargePersistentEpidemic1999; @hillemanRecoveryNewAgent1954; @ishikoSpreadEpidemicKeratoconjunctivitis2009; @kajonAdenovirusTypeRespiratory2018; @kandelOutbreakAdenovirusType2010; @kolavic-grayLargeEpidemicAdenovirus2002; @leeComprehensiveSerotypingEpidemiology2010; @linTwoDecadeSurvey2004; @liSwimmingPoolassociatedOutbreak2018; @mcneillEpidemicSpreadAdenovirus2000; @roweIsolationCytopathogenicAgent1953; @topControlAdenovirusAcute1975; @tulloOutbreakAdenovirusType1980; @vanderveenPatternsInfectionsAdenovirus1969; @yeungCharacterizationCulturepositiveAdenovirus2009; @zhangSurveyRecentAdenoviral2019]. A recent study suggests that this type consists of two phylogroups evolving at different rates [@gonzalezGenomicCharacterizationHuman2019].

### HAdV-B55

B55 is increasingly notorious. The first B55 samples originated during a 20 year period starting in 1965 at the 302nd Hospital in Beijing, China [@liGeneticRelationshipThirteen1991]. In March 1969, the first military outbreak of B55 occurred at the Sant Climent Sescebes camp of Alt Empordà, Girona, Catalonia, Spain [@hierholzerOCCURRENCERESPIRATORYILLNESS1974]. The virus continues spreading throughout the world  [@caoEmergenceCommunityAcquiredAdenovirus2014; @chmielewiczRespiratoryDiseaseCaused2005; @hierholzerOCCURRENCERESPIRATORYILLNESS1974; @jingHouseholdTransmissionHuman2019; @kajonMolecularEpidemiologyAdenovirus1996; @kajonMolecularSerologicalCharacterization2013; @kajonOutbreakFebrileRespiratory2010; @lafolieSeverePneumoniaAssociated2016; @liGeneticRelationshipThirteen1991; @liOutbreakAcuteRespiratory2014; @luEpidemiologyHumanAdenovirus2014; @salamaOutbreakAdenovirusType2016; @sunEmergentSevereAcute2014; @yiClusterAdenovirusType2017; @yooFebrileRespiratoryIllness2017; @zhuOutbreakAcuteRespiratory2009]. B55 is a re-emergent respiratory pathogen with a B14 genomic backbone and a recombinant hexon partially derived from B11 [@walshComputationalAnalysisIdentifies2010]. A new typing scheme that includes genomic analysis corrected it’s previous misidentification as B11a due to limitations associated with serological assays with respect to recombination effects [@walshComputationalAnalysisIdentifies2010].

## Biology (Summary)

The *Adenoviridae* family includes five genera that infect a wide range of hosts and cell types. The *Mastadenovirus* genus includes seven species (A-G) with over 103 types that infect mammalian hosts [@HadvwgGmuEdu; @lionAdenovirusInfectionsImmunocompetent2014]. They are globally distributed, causing sporadic outbreaks in densely populated regions and close-living quarters. Carriers may expose others via aerosol or fecal-oral transmission, potentially asymptomatically [@lynchAdenovirusEpidemiologyGlobal2016]. Symptoms range from acute respiratory disease to organ failure, depending on the viral species and host immune strength [@lynchAdenovirusEpidemiologyGlobal2016]. Accordingly, individuals with developing or weakened immune systems account for most outbreak deaths. The *Adenoviridae* are class I, linear, double-stranded DNA viruses [@baltimoreExpressionAnimalVirus1971]. The nonenveloped icosahedral nucleocapsid consists of hexon and penton capsomers forming the faces and vertexes respectively. The host coxsackie adenovirus receptor recognizes the fiber knob while the penton RGD motif induces structural changes to gain entry into the host [@petterssonEncountersAdenovirus2019].

## Prior Work

Previous adenovirus research has included molecular clock analysis to characterize outbreak samples. A common protocol involves a preliminary pseudo-statistical test of the strict molecular clock hypothesis followed by Markov chain Monte Carlo (MCMC) simulation to estimate model parameters and unknown uncertainties. Tracing each MCMC run evaluates convergence and determines whether the estimates sample size (ESS) of each parameter of interest is acceptable. A model selection step then compares runs with different clock hypotheses, coalescent models, and substitution models. The result is an estimation of model parameters with confidence intervals, such as divergence times and substitution rates. Recent adenovirus research has featured similar protocols. One paper analyzed B55 isolates between June 2009 and January 2012 in Chongqing, China [@luEpidemiologyHumanAdenovirus2014] and another paper analyzed E4 isolates obtained between 1953 and 2015 in the United States and Japan [@gonzalezGenomicCharacterizationHuman2019].

## Proposed Work

- Hypothesis: the genes that encode surface proteins such evolve faster than those encoding core machinery.
- Hypothesis: recombination is an evolutionary force for the emergence of new genotypes.
- Build a pipeline that automates the construction of data set representative of a measurably evolving population.
- Build a pipeline that automates molecular clock analysis.
- Build and publish a package that normalizes geographic data.
- Perform phylogeographic and molecular clock analysis for selected adenovirus genotypes: HAdV-E4 & HAdV-B55

# Methods

This section documents an executable pipeline that constructs data sets representative of measurably evolving populations and performs molecular clock analysis. The solution relies on **Snakemake**, which is a portable, rule-based workflow engine [@kosterSnakemakeScalableBioinformatics2012]. Each rule defines a step in the workflow. A rule can also configure its execution environment via the Conda package manager. The engine automatically infers the workflow path and parallelization based on input dependencies, creating a directed acyclic graph [@kosterSnakemakeScalableBioinformatics2012]. Accordingly, the engine guarantees the reproducibility of each step.

## Pipeline

### Phase 1: Data Set Generation

![phase-1: measurably evolving population](phase-1.svg)

The first phase uses a query to generate a set of timestamped, homologous sequences. The execution has two initial paths to process genes and genomes separately. Both paths use different strategies to guarantee full alignment coverage of the query. For genes, the first rule runs the **BLAST+** `blastn` program to perform local alignment and generate a library of sequences and the next rule runs the **FASTA** `glsearch36` program to perform global-local alignment [@camachoBLASTArchitectureApplications2009; @pearsonFASTASearchPrograms2014]. For genomes, the first rule runs the **BLAST+** `blastdbcmd` program to subset the BLAST database by sequence length, accepting those within a percentage deviation of expected size and the next rule runs the `nucmer` and `show-coords` programs of the **MUMmer** genome alignment suite [@marcaisMUMmer4FastVersatile2018]. Each path generates a report of query coverage identity scores.

The next rule defines a Python program that extracts the accessions from the report. It uses an **Entrez Direct** binding of the `esummary` utility to query **GenBank** and download a JSON file of the metadata [@kansEntrezDirectEutilities2020]. The final rule processes the JSON file and query coverage identity report. It extracts the "collection_date" qualifier and attempts to normalize it into an ISO-8601 string from a list of formats. The rule then accepts sequences based on an identity threshold and successful date extraction.

### Phase 2: Phylogenetic Analyses

The next phase performs phylogenetic analyses on the generated data set. The initial rule runs the `mafft` program to generate a multiple sequence alignment. This program calculates a fast Fourier transform to cluster and progressively align the sequences [@katohMAFFTNovelMethod2002]. It also automatically sets the optimal program execution mode based on input size and reverse-complements any sequence if necessary [@katohMAFFTNovelMethod2002]. Downstream rules process the resulting FASTA file.

Tree inference occurs in a rule that runs the `iqtree` program to infer a maximum-likelihood tree [@nguyenIQTREEFastEffective2015]. The program uses `ModelFinder` to calculate the best substitution model based on the Bayesian information criterion [@kalyaanamoorthyModelFinderFastModel2017]. The "-alrt" and "-bb" flags set the number of bootstrap replicates for the approximate likelihood ratio test of branches and branch support and the "-bnni" flag activates nearest neighbor interchange search optimization strategy [@anisimovaSurveyBranchSupport2011; @hoangUFBoot2ImprovingUltrafast2018]. The program outputs a log file and exports the tree in Newick format.

Recombination analysis occurs in a rule that runs the **Gubbins** program. It infers a phylogeny while identifying and accounting for recombination [@croucherRapidPhylogeneticAnalysis2015]. This step is necessary since recombination obscures temporal analysis [@rambautExploringTemporalStructure2016]. Accordingly, the output is a tree based on the clonal frame, the sequence with recombinant regions removed [@milkmanMolecularEvolutionEscherichia1990].

The pipeline also includes a rule to output **BEAST** model files [@suchardBayesianPhylogeneticPhylodynamic2018]. The rule executes a custom Python script that transforms the multiple sequence alignment and `ModelTest` result into an XML file that also specifies the sequence and sampling date for each taxon. The script maps the inferred substitution model to a **BEAST** compatible format. Clock models include the strict clock, relaxed clock with lognormal distribution, and relaxed clock with exponential distribution [@drummondRelaxedPhylogeneticsDating2006]. Tree priors include the Constant Size and Exponential Growth coalescent models [@griffithsSamplingTheoryNeutral1994; @kingmanCoalescent1982]. Also, the script includes specifications to run maximum likelihood estimation using the path sampling stepping stone method with a Beta path step distribution [@baeleAccurateModelSelection2012; @baeleImprovingAccuracyDemographic2012]. 

## Geographic Data

The geography of an isolate helps to reconstruct outbreak scenarios and is valuable for phylogeographic methods. It is also a problematic field to parse since it is necessary to normalize place names and countries. Preliminary research has yielded a new R package inspired by **lubridate** to this end. The **lubricity** package uses admin code data from the **GeoNames** project to clean such data. The package parses data frame columns with a list of user-specified, named, regular expressions. The name of each capture group corresponds to an administrative division. Matches to higher divisions limit potential matches to lower ones. This avoids issues when names are the same across different countries, etc...

# Results (Preliminary)

The sequence variation plot highlights mutations inferred from the multiple sequence alignment. These include transitions, transversions, insertions, and deletions. The gray boxes also indicate potential imports into the clonal frame via recombination.

BEAST generated trees for the cross-product of clock and population coalescent models. The "rex", "rln", and "str" labels are abbreviations for the relaxed exponential, relaxed lognormal, and strict clock models. The "con" and "exp" labels are abbreviations for the constant and exponential population coalescent models. Note that the strict, constant population model is the null hypothesis. Analysis considers the first 10% of values as burn-in and therefore discards those states.

The model result tables highlight satisfactory ones in bold. These are models with ESS values greater than or equal to 200 for the joint, prior, likelihood, age(root), and mean(rate) parameters. These tables also display the Bayes factor according to the path-sampling stepping stone method and median value for each parameter. Some models failed to exhibit a sufficient ESS value despite the greater Bayes factor, rendering them unsuitable for further analysis and interpretation. This suggests a need for increasing the simulation chain length. The log trace figures shows the post burn-in value of each parameter as the simulation progressed. The logs demonstrate how the values converged. Only the age(root) a greater variation with rare, large fluctuations at some points. For the HAdV-B55 study, the "acc" column corresponds to sequence accession.

Each chronogram corresponds to the satisfactory model with targets Bayes factor. The node error bars indicate the 95%-CI height (age) value. Most node heights exhibited small confidence intervals except for an extremely large one at the root, which reflects the variance observed in the log trace figure. This may be due to insufficient chain length, sequencing error, recombination, lack of samples, or local, non-clock behavior.

The HAdV-E4 analysis yielded a similar chronogram topology with clear separation of both phylogroups. Analysis also suggests a similar substitution rate of the same order at 4.09e-05 s/y/y. However, the age of the root was much younger compared to that in the paper: 1682.5797 vs 602. The HAdV-B55 analysis yielded an estimate substitution rate of 1.91e-05 s/y/y and TMRCA of 1532.696, comparable to the HAdV-4 result of this study.

## HAdV-B55

### Genomic Sequence Variation

```{r}
acc <- "FJ643676.1"

meta <- 
  file.path(root, acc, "meta.json") %>%
  read_file() %>%
  jqr::jq(".result | del(.uids) | map([.accessionversion, .title, .taxid, .organism, .subtype, .subname]) |.[]") %>% 
  textConnection() %>% 
  jsonlite::stream_in(simplifyDataFrame = T, verbose = F) %>%
  setNames(c("accver", "title", "taxid", "organism", "subtype", "subname")) %>%
  bind_cols(
    .,
    bind_rows(
      apply(., 1, function(row) {
        key <- str_split(row["subtype"], "\\|")[[1]]
        val <- str_split(row["subname"], "\\|")[[1]]
        data.frame(as.list(setNames(val, key)), stringsAsFactors = F)
      })
    )
  )

regx_list <- list(
  list(pattern = "(.+)", names = c("cn")),
  list(pattern = "(.+):\\s*(.+)", names = c("cn", "ac1")),
  list(pattern = "(.+):\\s*(.+)", names = c("cn", "ac2")),
  list(pattern = "(.+):\\s*(.+),", names = c("cn", "ac1")),
  list(pattern = "(.+):\\s*(.+),", names = c("cn", "ac2")),
  list(pattern = "(.+):\\s*.+,\\s*(.+)", names = c("cn", "ac1")),
  list(pattern = "(.+):\\s*.+,\\s*(.+)", names = c("cn", "ac2")),
  list(pattern = "(.+):\\s*(.+)\\s*,\\s*(.+)", names = c("cn", "ac1", "ac2")),
  list(pattern = "(.+):\\s*(.+)\\s*,\\s*(.+)", names = c("cn", "ac2", "ac1"))
)

msa <- 
  file.path(root, acc, "msa-1.fna") %>%
  ape::read.dna(format = "fasta", as.character = T) %>%
  toupper()

meta <-
  lapply(rownames(msa), function(ele)
    setNames(
      str_match(ele, "(.+)_(\\d{4}-\\d{2}-\\d{2})$")[1, ],
      c("label", "accver", "date")
    )
  ) %>%
  map_df(bind_rows) %>%
  merge(meta, by = "accver", sort = F) %>%
  bind_cols(bind_rows(lapply(.$country, process_regx_list, regx_list = regx_list))) %>%
  mutate(relabel = str_c(accver, date, cn, taxid, sep = "|")) %>%
  select(label, relabel, everything())

rownames(msa) <- recode(rownames(msa), !!!setNames(meta$relabel, meta$label))

tree <- file.path(root, acc, "gub.final_tree.tre") %>% ape::read.tree()
tree$tip.label <- recode(tree$tip.label, !!!setNames(meta$relabel, meta$label))
tree <- ape::ladderize(tree)
lvls <- tree$tip.label[tree$edge[tree$edge[,2] <= length(tree$tip.label), 2]]

vcf <- 
  bind_rows(merge(call_snp(msa), muts), mutate(call_ind(msa), pos = start)) %>%
  mutate(
    accver = factor(rownames(msa)[idx], levels = lvls), 
    call = factor(call, levels = c("trs", "trv", "sim", "dis", "ins", "del"))
  )

len <- ncol(msa)
colors <- list(trs = "blue", trv = "red", sim = "purple", dis = "orange", ins = "black", del = "black")
shapes <- list(trs = 124, trv = 124, sim = 124, dis = 124, ins = 6, del = 2)

recom <-
  file.path(root, acc, "gub.recombination_predictions.gff") %>%
  ape::read.gff() %>%
  apply(1, function(row) {
    read_delim(str_split(row["attributes"], ";")[[1]], "=", col_names = c("key", "val"), col_types = "cc") %>%
      filter(key == "taxa") %>% pull(val) %>% str_trim() %>% str_split("\\s+") %>% unlist() %>%
      enframe(name = NULL, value = "accver") %>%
      mutate(start = row["start"], end = row["end"])
  }) %>%
  bind_rows() %>%
  mutate(accver = factor(recode(accver, !!!setNames(meta$relabel, meta$label)), levels = lvls)) %>%
  mutate_at(c("start", "end"), as.integer)
```

```{r}
ggplot() +
  geom_point(data = vcf, aes(pos, accver, color = call, shape = call)) +
  geom_rect(
    data = recom,
    aes(
      xmin = start - 0.5, xmax = end + 0.5, 
      ymin = as.numeric(accver) - 0.5, ymax = as.numeric(accver) + 0.5
    ), 
    fill = "lightgrey", color = "black", alpha = 0.25
  ) +
  scale_y_discrete(limits = lvls) +
  scale_color_manual(values = colors) +
  scale_shape_manual(values = shapes) +
  xlim(1, len) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    text = element_text(family = "mono"),
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    axis.text.y = element_text(hjust = 0)
  )
```

### Model Evaluation

```{r}
df.bf <-
  list.files(root, pattern = ".mle.result.log", recursive = T, full.names = T) %>%
  lapply(parse_mle_report) %>%
  bind_rows() %>%
  mutate(acc = basename(dirname(path)), name = basename(path)) %>%
  separate(name, c("clock", "coal"), sep = "[-.]", extra = "drop") %>%
  group_by(acc) %>%
	mutate(
	  `BF (PS)` = PS - PS[which(clock == "str" & coal == "con")],
		`BF (SS)` = SS - SS[which(clock == "str" & coal == "con")]
	) %>%
  ungroup() %>%
  arrange(acc, desc(`BF (PS)`))
```

```{r}
logs <-
  list.files(root, pattern = "\\w+-\\w+.log", recursive = T, full.names = T) %>%
  grep("(rex|rln|str)-(con|exp).log", ., value = T) %>%
  grep("HAdV-4", ., invert = T, value = T)

pburn <- 0.1

df.ess <-
  mclapply(logs, function(path) {
    df <- parse_beast_log(path)
    interval <- df$state[2] - df$state[1]
    select(df, joint, prior, likelihood, age.root., meanRate, coalescent) %>%
      apply(2, function(ele) calc_ess(remove_burn_in(ele, pburn), interval)) %>%
      c(path = path)
  }) %>%
  map_df(bind_rows) %>%
  mutate(acc = basename(dirname(path)), path = basename(path)) %>%
  select(acc, path, everything()) %>%
  separate(path, c("clock", "coal"), sep = "[-.]", extra = "drop") %>%
  mutate_at(4:ncol(.), as.double)

df.est <- 
  mclapply(logs, function(path) {
    parse_beast_log(path) %>%
      remove_burn_ins(pburn) %>%
      apply(2, median) %>%
      c(path = path)
  }) %>%
  map_df(bind_rows) %>%
  mutate(acc = basename(dirname(path)), path = basename(path)) %>%
  select(acc, path, everything()) %>%
  separate(path, c("clock", "coal"), sep = "[-.]", extra = "drop") %>%
  mutate_at(4:ncol(.), as.double)
```

```{r}
df.res <-
  merge(df.est, df.ess, by = c("acc", "clock", "coal"), suffixes = c(".est", ".ess")) %>%
  merge(df.bf, ., sort = F) 

keys <- paste0(c("joint", "prior", "likelihood", "age.root.", "meanRate", "coalescent"), ".ess")
idx <- (select(df.res, all_of(keys)) >= 200) %>% apply(1, all) %>% which()
```

```{r}
select(df.res, acc, `BF (PS)`, clock, coal, starts_with(c("age.root.", "meanRate")), -path, -state) %>%
  knitr::kable("html", col.names = c("acc", "PS", "clock", "coal", c("est" ,"ess", "est" ,"ess"))) %>%
  add_header_above(c(" " = 1, "BF", "model" = 2, "age(root)" = 2, "mean(rate)" = 2)) %>%
  row_spec(idx, bold = T)
```

```{r}
best <- df.res[idx, ] %>% .[.$acc == acc, ] %>% head(1) %>% with(paste0(clock, "-", coal))
```

### Chronogram (`r best`)

```{r, warning=FALSE}
tree <-
  paste0(best, ".mcc.tree") %>%
  file.path(root, acc, .) %>%
  read.beast()
tree@phylo$tip.label <- recode(tree@phylo$tip.label, !!!setNames(meta$relabel, meta$label))
plot_chronogram(tree)
```

```{r}
paste0(best, ".log") %>%
  file.path(root, acc, .) %>%
  parse_beast_log() %>%
  select(state, joint, prior, likelihood, age.root., meanRate, coalescent) %>%
  filter(state >= max(state) * pburn) %>%
  pivot_longer(2:ncol(.), names_to = "statistic") %>%
  ggplot(aes(state, value)) + 
    geom_path() + 
    facet_wrap(~ statistic, ncol = 1, scales = "free_y", strip.position = "right") +
    theme_minimal()
```

## HAdV-E4

### Genomic Sequence Variation

```{bash, script}
mkdir -p exp/HAdV-4 && cd exp/HAdV-4

# efetch
[ ! -f rec.fna ] && awk 'NR > 1 { print $1; }' meta.tsv | tr '\n' ',' | xargs efetch -db nuccore -format fasta -id > rec.fna

# meta
awk -F '\t' 'NR > 1 { printf "/^>/ s/%s/%s_%s_%s/\n", $1, $1, $4, $5; }' meta.tsv > meta.sed

# mafft
[ ! -f msa-1.fna ] && 
sed '/^>/ s/ .*//' rec.fna | sed -f meta.sed | mafft --auto --adjustdirection --thread -1 - > msa-1.fna 2> msa-1.log

# gubbins
sed '/^[^>]/ s/[^acgtACGT-]/N/g' msa-1.fna > msa-2.fna
[ ! -f gub.log ] && run_gubbins.py --prefix gub --iterations 1000 --threads 16 msa-2.fna > gub.log
bedtools maskfasta -fi msa-1.fna -fo msa-2.fna -bed gub.recombination_predictions.gff

# iqtree
[ ! -f phy.log ] && iqtree -s msa-2.fna -pre phy -alrt 1000 -bb 1000 -bnni -nt 16 > /dev/null 2> /dev/null

# beautify
line=( $(
  grep -v WARNING phy.log | awk '{{$1=$1;}} //1;' | \
    grep -A 131 '1 JC' | grep -v +R | sort -n -k 7 | head -n 1 | \
    sed -e 's/+G4/+G/g' -e 's/+F//g'
) );
for clock in "rex" "rln" "str"; do
  for coal in "con" "exp"; do
    echo ../../../../workflow/scripts/beautify.py \
      -dregex "_(\d{4})_" -dformat "%Y" -len_mcmc 100000000 -stem "$clock-$coal" \
      msa-2.fna ../../../../workflow/templates "${line[1]}" "$clock" "$coal" > "$clock-$coal.xml"
  done
done
```

```{r}
subroot<- "HAdV-4"

msa <- 
  file.path(root, subroot, "msa-1.fna") %>%
  ape::read.dna(format = "fasta", as.character = T) %>%
  toupper()

tree <- file.path(root, subroot, "gub.final_tree.tre") %>% ape::read.tree()
tree <- ape::ladderize(tree)
lvls <- tree$tip.label[tree$edge[tree$edge[,2] <= length(tree$tip.label), 2]]

vcf <- 
  bind_rows(merge(call_snp(msa), muts), mutate(call_ind(msa), pos = start)) %>%
  mutate(
    accver = factor(rownames(msa)[idx], levels = lvls), 
    call = factor(call, levels = c("trs", "trv", "sim", "dis", "ins", "del"))
  )

len <- ncol(msa)
colors <- list(trs = "blue", trv = "red", sim = "purple", dis = "orange", ins = "black", del = "black")
shapes <- list(trs = 124, trv = 124, sim = 124, dis = 124, ins = 6, del = 2)

recom <-
  file.path(root, subroot, "gub.recombination_predictions.gff") %>%
  ape::read.gff() %>%
  apply(1, function(row) {
    read_delim(str_split(row["attributes"], ";")[[1]], "=", col_names = c("key", "val"), col_types = "cc") %>%
      filter(key == "taxa") %>% pull(val) %>% str_trim() %>% str_split("\\s+") %>% unlist() %>%
      enframe(name = NULL, value = "accver") %>%
      mutate(start = row["start"], end = row["end"])
  }) %>%
  bind_rows() %>%
  mutate_at("accver", factor, levels = lvls) %>%
  mutate_at(c("start", "end"), as.integer)
```

```{r}
ggplot() +
  geom_point(data = vcf, aes(pos, accver, color = call, shape = call)) +
  geom_rect(
    data = recom,
    aes(
      xmin = start - 0.5, xmax = end + 0.5, 
      ymin = as.numeric(accver) - 0.5, ymax = as.numeric(accver) + 0.5
    ), 
    fill = "lightgrey", color = "black", alpha = 0.25
  ) +
  scale_y_discrete(limits = lvls) +
  scale_color_manual(values = colors) +
  scale_shape_manual(values = shapes) +
  xlim(1, len) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    text = element_text(family = "mono"),
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    axis.text.y = element_text(hjust = 0)
  )
```

### Model Evaluation

```{r}
df.bf <-
  file.path(root, subroot) %>%
  list.files(pattern = ".mle.result.log", recursive = T, full.names = T) %>%
  lapply(parse_mle_report) %>%
  bind_rows() %>%
  mutate(acc = basename(dirname(path)), name = basename(path)) %>%
  separate(name, c("clock", "coal"), sep = "[-.]", extra = "drop") %>%
  group_by(acc) %>%
	mutate(
	  `BF (PS)` = PS - PS[which(clock == "str" & coal == "con")],
		`BF (SS)` = SS - SS[which(clock == "str" & coal == "con")]
	) %>%
  ungroup() %>%
  arrange(acc, desc(`BF (PS)`))

logs <-
  file.path(root, subroot) %>%
  list.files(pattern = "\\w+-\\w+.log", recursive = T, full.names = T) %>%
  grep("(rex|rln|str)-(con|exp).log", ., value = T)

pburn <- 0.1

df.ess <-
  mclapply(logs, function(path) {
    df <- parse_beast_log(path)
    interval <- df$state[2] - df$state[1]
    select(df, joint, prior, likelihood, age.root., meanRate, coalescent) %>%
      apply(2, function(ele) calc_ess(remove_burn_in(ele, pburn), interval)) %>%
      c(path = path)
  }) %>%
  map_df(bind_rows) %>%
  mutate(path = basename(path)) %>%
  select(path, everything()) %>%
  separate(path, c("clock", "coal"), sep = "[-.]", extra = "drop") %>%
  mutate_at(3:ncol(.), as.double)

df.est <-
  mclapply(logs, function(path) {
    parse_beast_log(path) %>%
      remove_burn_ins(pburn) %>%
      apply(2, median) %>%
      c(path = path)
  }) %>%
  map_df(bind_rows) %>%
  mutate(path = basename(path)) %>%
  select(path, everything()) %>%
  separate(path, c("clock", "coal"), sep = "[-.]", extra = "drop") %>%
  mutate_at(4:ncol(.), as.double)
```

```{r}
df.res <-
  merge(df.est, df.ess, by = c("clock", "coal"), suffixes = c(".est", ".ess")) %>%
  merge(df.bf, ., sort = F)

keys <- paste0(c("joint", "prior", "likelihood", "age.root.", "meanRate", "coalescent"), ".ess")
idx <- (select(df.res, all_of(keys)) >= 200) %>% apply(1, all) %>% which()
```

```{r}
select(df.res, `BF (PS)`, clock, coal, starts_with(c("age.root.", "meanRate")), -path, -state) %>%
  knitr::kable("html", col.names = c("PS", "clock", "coal", c("est" ,"ess", "est" ,"ess"))) %>%
  add_header_above(c("BF" = 1, "model" = 2, "age(root)" = 2, "mean(rate)" = 2)) %>%
  row_spec(idx, bold = T)
```

```{r}
best <- df.res[idx, ] %>% head(1) %>% with(paste0(clock, "-", coal))
```

### Chronogram (`r best`)

```{r, warning=FALSE, message=FALSE}
tree <-
  paste0(best, ".mcc.tree") %>%
  file.path(root, subroot, .) %>%
  read.beast()

phylogroup <- 
  enframe(tree@phylo$tip.label, name = NULL, value = "label") %>% 
  separate(label, c("accver", "year", "country"), sep = "_", remove = F) %>% 
  merge(read_tsv("exp/HAdV-4/meta.tsv")) %>% 
  with(split(label, phylogroup))

tip.date <- str_match(tree@phylo$tip.label, "_(\\d{4})_")[, 2]
groupOTU(tree, phylogroup, "phylogroup") %>%
  ggtree(mrsd = paste0(min(tip.date), "-01-01")) +
  aes(color = phylogroup) +
  geom_tiplab(linesize = 1, align = T, color = "black") +
  geom_range("height_0.95_HPD") +
  theme_tree2() +
  theme(
    legend.position = "bottom",
    panel.grid.major.x = element_line(color="black", size = .25),
    panel.grid.minor.x = element_line(color="grey", size = .25)
  )
```

```{r}
paste0(best, ".log") %>%
  file.path(root, acc, .) %>%
  parse_beast_log() %>%
  select(state, joint, prior, likelihood, age.root., meanRate, coalescent) %>%
  filter(state >= max(state) * pburn) %>%
  pivot_longer(2:ncol(.), names_to = "statistic") %>%
  ggplot(aes(state, value)) + 
    geom_path() + 
    facet_wrap(~ statistic, ncol = 1, scales = "free_y", strip.position = "right") +
    theme_minimal()
```

# Timeline

```{r}
gannt <- 
read_tsv("
project	class	start	finish
Characterization of an Adenovirus Clinical Isolate	BINF703	August 27, 2018	December 8, 2018
Minimum Set Cover Problem Pipeline for Multi K-mer Barcoding	BINF703	August 27, 2018	December 8, 2018
Analysis of Three Adenovirus Outbreaks and an Indel-Calling Algorithm	BINF703	January 22, 2019	May 6, 2019
Doctoral Dissertation Proposal (1)	BINF998	January 22, 2019	May 6, 2019
Doctoral Dissertation Proposal (2)	BINF998	May 20, 2019	August 7, 2019
Doctoral Dissertation Proposal (3)	BINF998	August 26, 2019	December 7, 2019
US-Japan Joint Panel Conference on Cholera	BINF998	December 10, 2019	December 13, 2019
Doctoral Dissertation Proposal (4)	BINF998	January 21, 2020	May 11, 2020
Chapter 1: Adenovirus History	BINF999	June 1, 2020	June 30, 2020
Chapter 1: Adenovirus Biology	BINF999	June 1, 2020	June 30, 2020
Chapter 1: HAdV-E4 History	BINF999	July 1, 2020	August 31, 2020
Chapter 1: HAdV-E4 Biology	BINF999	July 1, 2020	August 31, 2020
Chapter 1: HAdV-B55 History	BINF999	July 1, 2020	August 31, 2020
Chapter 1: HAdV-B55 Biology	BINF999	July 1, 2020	August 31, 2020
Chapter 2: Evolutionary Forces on Sequences	BINF999	September 1, 2020	September 30, 2020
Chapter 2: Substitution Model Testing	BINF999	September 1, 2020	September 30, 2020
Chapter 2: Sequence Homology Search Methods	BINF999	September 1, 2020	September 30, 2020
Chapter 2: Phylogenetic Methods	BINF999	September 1, 2020	September 30, 2020
Chapter 2: Molecular Clock Models and Methods	BINF999	October 1, 2020	November 30, 2020
Chapter 2: Clock Signal Testing	BINF999	October 1, 2020	November 30, 2020
Chapter 2: Molecular Clock Model Testing	BINF999	October 1, 2020	November 30, 2020
Chapter 2: Data Acquisition/Normalization	BINF999	October 1, 2020	November 30, 2020
Chapter 2: Reproducibility	BINF999	October 1, 2020	November 30, 2020
Chapter 2: Pipeline	BINF999	December 1, 2020	January 31, 2021
Chapter 3: HAdV-E4	BINF999	February 1, 2021	April 30, 2021
Chapter 4: HAdV-B55	BINF999	February 1, 2021	April 30, 2021
Chapter 5: Conclusions	BINF999	May 1, 2021	July 31, 2021
") %>%
  mutate_at(c("start", "finish"), as.POSIXct, format = "%b %d, %Y") %>%
  arrange(desc(start))

gannt$project <- factor(gannt$project, levels = gannt$project)
ggplot() +
  geom_rect(
    data = gannt,
    aes(
      xmin = start, xmax = finish, 
      ymin = as.numeric(project) - 0.5, ymax = as.numeric(project) + 0.5,
      fill = class
    )
  ) +
  scale_x_datetime(labels = scales::date_format("%b %Y"), date_breaks = "1 month") +
  scale_y_discrete(limits = gannt$project) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))

arrange(gannt, start) %>% knitr::kable("html")
```

# References
